### PR TITLE
Reland "[spirv] Switch to use common target description"

### PR DESCRIPTION
**Pull request by antiagainst**
This relands https://github.com/iree-org/iree/pull/17623.

This commit switches SPIR-V side to use the common `#iree_gpu.target` to describe the GPU characteristics. With it we can now remove the ad-hoc Vulkan attributes and dialects and unify how GPU are described across various GPU compiler backends in IREE.

SPIR-V has some additional requirements that we need to account for:

We have many vendors and APIs to handle there so this commit adds various AMD/ARM/NVIDIA/Qualcomm targets for
development purposes so that we can specify them with a shorthand.

In order to be extensible, leverage the `feature` field in `#iree_gpu.target` to specify additional capabilities with `cap:` prefix and extensions with `ext:` prefix. We also use the `feature` field to specify what SPIR-V version to target with the `spirv:v1.x` format.

Right now the `SPIRVConvertGPUTarget` pass is
invoked immediately before configuration. This is to stage the changes. As a next step we need to move
it immediately before `ConvertToSPIRV` pass.

`--iree-vulkan-target-env` is dropped given now we removed the whole Vulkan dialect and cannot control with a `#vk.target_env` attribute anymore.

The default `--iree-vulkan-target-triple` now becomes `vp_android_baseline_2022`, which is a a good lowest common denominator to guarantee the generated SPIR-V is widely accepted. We are not considering SwiftShader now anymore like previously due to testing purposes.

The `--iree-vulkan-target-triple` should be renamed given it's not a triple anymore--that will happen later together with other GPU backends (i.e., cuda/hip) to be consistent.

In order to support cooperative matrix conversion, we added `WMMA_F16_16x16x16_F16`. For NVIDIA GPUs
we are abusing it right now without considering the concrete explicit layout--that is fine given in Vulkan they are opaque anyway. But this need to be fixed if we are targeting WMMA in CUDA.

We now contruct a `#iree_gpu.target` to specify
the target to drive SPIR-V CodeGen.

Progress towards https://github.com/iree-org/iree/issues/16341

ci-extra: test_nvidia_gpu,test_nvidia_a100,test_amd_mi250, build_test_all_macos_arm64,build_and_test_android,test_on_moto-edge-x30

### CODE REVIEW

**Comment by MacDue** [comment]
(typo in PR title `Raland` -> `Reland`)

**Comment by antiagainst** [comment]
@ScottTodd @kuhar this should be good now--android tests and benchmarks are fixed.

File: .git-blame-ignore-revs [hunk 1]
```diff
@@ -12,3 +12,7 @@ be24f0258a520a48555c9baec9d2f737ba1c2ca0
 
 # Switch compiler to LLVM/MLIR formatting style (#14181)
 3b652d46e86cd82cc81922a964327b12dc914428
+
+# Land and revert using #iree_gpu.target for SPIR-V (#17623, #17698)
+7b9fb121ea1127a6d199904a0f2b0409573bc52d
+d792d2483a9d5d875a0f03b57c9d4f98e072c300
```
Comment (ScottTodd) [inline .git-blame-ignore-revs]: Eh, I'd keep these in `git blame` (revert changes to this file). We haven't added any other PR -> revert -> PR sequences to this file in this way before.

Comment (antiagainst) [reply .git-blame-ignore-revs]: Works for me! I just reverted the change there.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp [hunk 2]
```diff
@@ -576,6 +577,8 @@ static LogicalResult
 setVectorDistributionConfig(IREE::GPU::TargetAttr target,
                             mlir::FunctionOpInterface entryPoint,
                             Operation *computeOp) {
+  if (!isROCmBackend(target))
+    return failure();
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp]: This should have a comment. Why fail on just rocm? Is there some bug? Is something just not implemented yet?
(Comment should have enough info for someone in the future to determine when this `return failure()` is safe to remove)

**Review by ScottTodd — COMMENTED** [review]
One comment, otherwise LGTM

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp (hunk 2 above)
Comment (antiagainst) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp]: Good point. This pass is not yet supported for CUDA. Added comment to make it clear.

**Review by kuhar — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp (hunk 2 above)
Comment (ScottTodd) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp]: Ah, thanks. I missed the `!` character too :P

**Review by ScottTodd — APPROVED** [review]

**Comment by antiagainst** [comment]
Looks `test_amd_mi250` is just stuck there. That was fine in the previous patch already; so I'll just land this for now.

**Comment by ScottTodd** [comment]
> Looks `test_amd_mi250` is just stuck there. That was fine in the previous patch already; so I'll just land this for now.

That runner went offline for maintenance yeah.

**Pull request merged**